platform-sdk/docs/proposals/state-lifecycle/proposal.md

```diff
@@ -0,0 +1,86 @@
+# Lifecycle management for the State
+
+## Summary
+
+This proposal describes a possible implementation of state lifecycle management that could work equally well for both the Consensus Node and the Block Node.
+
+| Metadata           | Entities                                   |
+|--------------------|--------------------------------------------|
+| Designers          | [@imalygin](https://github.com/imalygin) |
+| Functional Impacts | Consensus Node, Block Node                 |
+
+## Purpose and Context
+
+There is an initiative to create self-contained State modules (`swirlds-state-api` and `swirlds-state-impl`) that would have minimal dependencies, and specifically, they should not depend on `swirlds-platform-core`. Part of this initiative is the implementation of a state lifecycle management mechanism that could be used by both the Consensus Node and the Block Node.
+
+### Requirements
+
+- Add an interface representing state lifecycle management and its implementation.
+- Update the current State API to be coherent with this State Lifecycle API.
+- Implement this set of APIs and migrate the existing code to use it.
+
+## Design & Architecture
+
+### Java classes
+
+- `com.swirlds.state.State` has the following lifecycle-related methods:
+  - `State copy()`: creates a copy of the state, which is mutable. The source of the copy becomes immutable.
+  - `void computeHash()`: computes the hash of the immutable state.
+  - `void createSnapshot(Path targetPath)`: creates a snapshot of the state at the specified path.
+  - `State loadSnapshot(Path sourcePath)`: loads the snapshot from the specified path.
```

Comment:
I think it can be static, at least for now. When we get to supporting multiple states from different state families in memory at the same time, we may decide to change it.

Reply:
As of now, this method is not static. Kelly is right, technically it can be static. However, it would be suboptimal because we'd have to move `MerkleTreeSnapshotReader` class to `swirlds-state-api`. As the name suggest, it knows about the implementation details of the `State` and therefore it doesn't belong to `swirlds-state-api`. 

I'd like to move this method to the `StateLifecycleManager` because we don't want this logic to require an instance of `State` object.

Reply:
Should this method still be here then? I agree, it should be in another class. `StateLifecycleManager` is good.

Reply:
It should not be there and it should be static and be a part of `StateLifecycleManagerImpl`